### DISCUSSION TITLE
Is styled components feasible for next app router?

### DISCUSSION THREAD

**Discussion by aomini** [vercel/next.js#50473] [App Router] [answered]
Since, the styled components make use of contexts I'm unable to do
```
export styledDiv = styled.div``
```
so, for every single components even without client side states I have to do 'use client'. Because of which I see page flickerings. How can we render styled components server side in app router?

**Comment by devjiwonchoi** [comment]
Have you configured the `layout.tsx` with [Server Side Rendering Styled-Components](https://styled-components.com/docs/advanced#server-side-rendering) using `ServerStyleSheet`?

**Reply by aomini** [reply]
Yes, I have used. Now, it's a less flickering but the question is even for a presentation component I cant create a component out of styled components. It says cannot use a client-only component with createContext.

**Reply by AndersDJohnson** [reply]
Same problem - get the following error when importing `styled-components`:
> TypeError: createContext only works in Client Components. Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/context-in-server-component

**Reply by aomini** [reply]
yeah. It only works that way :'(

**Comment by devjiwonchoi** [comment, marked as answer]
In `App Router`, use `styled-components API` to create a global registry component.

Example:
```js
// [lib/registry.tsx]

'use client';
 
import React, { useState } from 'react';
import { useServerInsertedHTML } from 'next/navigation';
import { ServerStyleSheet, StyleSheetManager } from 'styled-components';
 
export default function StyledComponentsRegistry({
  children,
}: {
  children: React.ReactNode;
}) {
  // Only create stylesheet once with lazy initial state
  // x-ref: https://reactjs.org/docs/hooks-reference.html#lazy-initial-state
  const [styledComponentsStyleSheet] = useState(() => new ServerStyleSheet());
 
  useServerInsertedHTML(() => {
    const styles = styledComponentsStyleSheet.getStyleElement();
    styledComponentsStyleSheet.instance.clearTag();
    return <>{styles}</>;
  });
 
  if (typeof window !== 'undefined') return <>{children}</>;
 
  return (
    <StyleSheetManager sheet={styledComponentsStyleSheet.instance}>
      {children}
    </StyleSheetManager>
  );
}
```

Then import `StyledComponentsRegistry` to your `root layout`.

Example:
```js
// [app/layout.tsx]

import StyledComponentsRegistry from './lib/registry';
 
export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html>
      <body>
        <StyledComponentsRegistry>{children}</StyledComponentsRegistry>
      </body>
    </html>
  );
}
```

By doing this, you'll be able to use the `export styledDiv = styled.div` format inside your `App Router` Components.

For more information, look [here](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#styled-components).

**Reply by ceddybi** [reply]
@devjiwonchoi thank you very much.

~~while this works in environment were javascript is available, how do you compile the styles on the server and then send them to the client in cases where the client has javascript disabled ?~~

**Reply by ceddybi** [reply]
my bad, the solution works 100% even if client disables javascript, i can still see my styles.

**Comment by SunStupic** [comment]
What about styled-component@5? Is it possible to use with nextjs app-router?
